### DISCUSSION TITLE
@emotion/react@11.13.3 CacheProvider issue

### DISCUSSION THREAD

**Discussion by simplecommerce** [garronej/tss-react#218] [Q&A] [answered]
Hi,

I have noticed a weird issue that I am having with my project when I update my packages to latest versions.
For some reason when updating @emotion/cache to 11.13.3, the CacheProvider doesn't seem to work with tss-react in SSR.
The version I have is: tss-react@4.9.13

If I substitute the CacheProvider with TssCacheProvider, it works again.
It was working fine before I did the update of @emotion/react@11.11.4 to @emotion/react@11.13.3.

I assume something must've changed in the package @emotion/react package.

Is the recommended usage to always use TssCacheProvider?

Update: Nevermind, TssProvider doesn't work, the page loads but I get client side issues with the className not matching.

It seems to cause the crash because I use:

> import { makeStyles } from "tss-react/mui";

Thanks!

**Comment by garronej** [comment]
Hello @simplecommerce,  
Thanks for reporting, can you create a reproduction repo?

**Reply by simplecommerce** [reply]
@garronej I will try to make a reproduction repo and get back to you, thanks!

**Reply by garronej** [reply]
Wait before you do, show me what your mareStyles looks like.

Do you use https://docs.tss-react.dev/nested-selectors#ssr anywhere?

**Reply by simplecommerce** [reply]
@garronej yes I am, but I did not have this issue prior to the update of @emotion/react

**Reply by garronej** [reply]
Yes, but it could very well have worked by coincidence.    
If you use nested selectors with the makeStyles API you have to provide a uniq id.  
Can you try and confirm it work?  

Anyway. My recomendation is to migrate from the makeStyles() API to [the modern API](https://docs.tss-react.dev/).  
The modern API does not have thoses caviats, it's much more less convoluted and you will experience significant performance gain.

**Reply by simplecommerce** [reply]
Yeah, during my migration from previous version to this one, I omitted using the modern API because I would've had to go through every page/component to restyle them using the new approach and I couldn't spend the time on that yet.

During my tests to try and find out what the issue was, I added a console.log here:
https://github.com/garronej/tss-react/blob/7851b98ea1b6a43a0fe4e6c4d82cf21f2ee91bef/src/makeStyles.tsx#L230

And the result would always be null when I use the latest version of @emotion/react vs the previous one I am currently using.

**Reply by simplecommerce** [reply]
Just an update, I did the following test to confirm your nested selectors theory.

```
import { makeStyles } from "tss-react/mui";

/**
 *
 */
const TestComponent = () => {
	const useStyles = makeStyles({
		name: "TestComponent",
	})({});

	useStyles({});

	return <div>test</div>;
};
```

When I render this component inside my CacheProvider, it crashes on the SSR.
If I comment the useStyles call, it works fine.

In my test, I commented my App code so that it doesn't render anything apart from this component in order to remove any doubt.

I will try to make a real repo to show you.

**Reply by garronej** [reply]
> And the result would always be null when I use the latest version of @emotion/react vs the previous one I am currently using.  

Okay this is bad.  

> Just an update, I did the following test to confirm your nested selectors theory.  

I don't know about that, this code dosen't make much sence to me.  

Let's forget about the nested selectors. This isn't it. Let's focus on the contextual cahe not beeing picked up in the latest version of @emotion/react.  
If you could reproduce that it would help a lot.

**Reply by garronej** [reply]
If it's confirmed I need to put a big warning everywhere until it's fixed.  

And the problematic package is @emotion/react not @emotion/cache.  

If you could determine the last version it works it would be great.

**Reply by simplecommerce** [reply]
Ok I was able to create a repo here: https://github.com/simplecommerce/vike-example-mui-emotion-cache/tree/simplecommerce/patch-51621
If you look at the file in `/src/app/PageShell.tsx`
I kept it simple.

You can see the live demo here: https://stackblitz.com/~/github.com/simplecommerce/vike-example-mui-emotion-cache

@emotion/react 11.13.3 (fails - currently installed)
tss-react 4.9.13

@emotion/react 11.12.0 (works)
tss-react 4.9.13

In the terminal you can simply do this to downgrade the version:

```
pnpm i @emotion/react@11.12.0
pnpm run dev
```

And after that, it should load fine.
Hope this helps, thanks!

**Comment by garronej** [comment, marked as answer]
Thank you for the detailed reproduction, @simplecommerce.

I can confirm that this is indeed a bug.

Although I can't resolve it directly on my end, I'll take the necessary steps to ensure the issue is addressed upstream.

For context, this is related to [the infamous dual package issue](https://github.com/garronej/vite-dual-package-repro-repo), which occurs when bundlers incorrectly determine which distribution of a module should be loaded.

Previously, a PR on Vite was required to fix a similar occurence of the same issue: https://github.com/vitejs/vite/pull/13370

Since it may take some time for the issue to be fully resolved, I recommend using Emotion 11.12 as a workaround in the meantime.

**Reply by garronej** [reply]
To be clear, this is a bug in Vite that affects tss-react.

**Reply by simplecommerce** [reply]
Thank you for all your help!

**Comment by garronej** [comment]
Just an update, I haven't forget about this and still plan on fixing it. I just need to allocate some hours to it.

**Reply by simplecommerce** [reply]
No worries! Thanks for the update!

**Comment by simplecommerce** [comment]
@garronej I had a quick question, I didn't want to open a new discussion just to ask this, it's related to the modern API.

I followed the instructions in the setup and for some reason the `theme` object is always undefined when I call `tss.create`.

At first I thought it was in my project, so I used the same repo in this discussion to test with version 11.12.0 of @emotion/react.

```
  const useStyles2 = tss.create(({ theme, props }) => {
    console.log("theme",theme)

    return {
        //If you don't need neither the theme nor any state or
        //props to describe your component style you can pass-in
        //an object instead of a callback.
        "root": {
            "color": "red"
        }
    }
  })
```

This always says undefined, I don't know why.

**Reply by garronej** [reply]
How do you import tss?  

If you import it with `import { tss } from "tss-react/mui"` you should have the mui theme.  
If you import lik `import { tss } from "tss-react";` the theme will be undefined.  

You can define what you want to be provided by creating:  

`src/tss.ts`
```ts
import { createTss } from "tss-react";

function useContext() {
    const myTheme = {
        primaryColor: "#32CD32", // This is LimeGreen in hex
    };
    
     const foo = 42;

    
    // You can return anything here, you decide what's the context.
    return { myTheme, foo };
}

export const { tss } = createTss({ useContext });

export const useStyles = tss.create({});
```

And then you can use it like:  

```ts
import { tss } from "tss"; // assuming you have baseUrl set to src

const useStyles= tss.create(({ myTheme, foo })=> ({/* ... */ }))
```

I thik this approach if very powerfull since it let you precisely define what is the context you'll need in your app for styles. It can go beyond just the mui theme object.  

By the wat the `tss-react/mui` is just an export of this:  

```ts
import { createTss } from "tss-react";
import { useTheme } from "@mui/material/styles";

function useContext() {
    const theme = useTheme();
    return { theme };
}

export const { tss } = createTss({ useContext });

export const useStyles = tss.create({});
```

**Reply by simplecommerce** [reply]
That was it, it was the import, I had to change it to `import { tss } from "tss-react/mui" `.

Sorry about that, thanks!!

**Reply by garronej** [reply]
On another note I saw in you code that you where using tss inside the body of your component.  
Do you know that you can do that for example:  

```tsx
import { useStyles } from "tss-react/mui";

function MyComponent(){

        const { css, cx, theme } = useStyles();
        
        return (
              <div className={css({ color: theme.colors.typography.main })} />
              <div className={cx(css({ /* ... */ }), props.classNames)} />
        );

}
```

It will be much more efficient.  

Also I noticed that you where looking to pass props, you'd do it this way:  

```tsx
import { tss } from "tss-react/mui";

type Props = {
     className?: string;
     foo: number;
     // some other props...
};

function MyComponent(props: Props) {

      const { className } = props;
      
      const { classes, cx } = useStyles({ props });
      
      return (
          <div 
            className={cx(classes.root, className)} // This enables you to make sure that the styles provided by the parent overwrite the internal styles.  
         > 
               {/* ... */}
          </div>
      );
}


const useStyles = tss
    .withName("MyComponent") // Help you debug with the inspect tool, for tracing back the CSS code.
    .withParams<{ props: Props; }>()  // Declare what params should be passed to useStyles ans what can be accessed below
    .create(({ theme, props }) => ({
        root: {
            color: props.foo
        }
    });

```  

I take the time to detail that because I lament the fact that most pepole just use tss as a quick way to migrate away from mui v4 makeStyles without taking the time to see what the sollution has to offers.  

I hope it helps a little bit.  

Best

**Reply by simplecommerce** [reply]
Thanks for taking the time to explain, I will definitely look into it more in depth! 
The repo was just an example to demonstrate the issue we discussed here.

Thanks again!

**Reply by garronej** [reply]
Lovely, I'll get back to you concerning the issue relative to emotion.  

PS: Reading my message again, I figured out that it came out as if I was acusing you of not reading the doc. Of course I didn't ment it that way. I just wanted to share some tips :)

**Reply by simplecommerce** [reply]
No worries, I didn't take it like that, appreciate the tips!

**Comment by garronej** [comment]
Resolution can be tracked here: https://github.com/emotion-js/emotion/issues/3267

**Comment by garronej** [comment]
The issue has been confirmed it's coming from Vike.  
Follow resolution: https://github.com/vikejs/vike/issues/1930
